dockerfile_parse/parser.py

```diff
@@ -15,72 +15,87 @@
 from six import string_types
+from six.moves import shlex_quote as quote
 
 from .constants import DOCKERFILE_FILENAME, COMMENT_INSTRUCTION
-from .util import (b2u, extract_labels_or_envs, get_key_val_dictionary,
+from .util import (b2u, extract_key_values, get_key_val_dictionary,
                    u2b, Context, WordSplitter)
 
-try:
-    # py3
-    from shlex import quote
-except ImportError:
-    from pipes import quote
-
 
 logger = logging.getLogger(__name__)
 
 
-class Labels(dict):
+class KeyValues(dict):
     """
-    A class for allowing direct write access to Dockerfile labels, e.g.:
+    Abstract base class for allowing direct write access to Dockerfile
+    instructions which result in a set of key value pairs.
 
-    parser.labels['label'] = 'value'
+    Subclasses must override the `parser_attr` value.
     """
+    parser_attr = None
 
-    def __init__(self, labels, parser):
-        super(Labels, self).__init__(labels)
+    def __init__(self, key_values, parser):
+        super(KeyValues, self).__init__(key_values)
         self.parser = parser
 
     def __delitem__(self, key):
-        super(Labels, self).__delitem__(key)
-        self.parser.labels = dict(self)
+        super(KeyValues, self).__delitem__(key)
+        setattr(self.parser, self.parser_attr, dict(self))
 
     def __setitem__(self, key, value):
-        super(Labels, self).__setitem__(key, value)
-        self.parser.labels = dict(self)
+        super(KeyValues, self).__setitem__(key, value)
+        setattr(self.parser, self.parser_attr, dict(self))
+
+    def __eq__(self, other):
+        if not isinstance(other, dict):
+            return False
+        return dict(self) == other
+
+    def __hash__(self):
+        return hash(json.dumps(self, separators=(',', ':'), sort_keys=True))
 
 
-class Envs(dict):
+class Labels(KeyValues):
+    """
+    A class for allowing direct write access to Dockerfile labels, e.g.:
+
+    parser.labels['label'] = 'value'
+    """
+    parser_attr = 'labels'
+
+
+class Envs(KeyValues):
     """
     A class for allowing direct write access to Dockerfile env. vars., e.g.:
 
     parser.envs['variable_name'] = 'value'
     """
+    parser_attr = 'envs'
 
-    def __init__(self, envs, parser):
-        super(Envs, self).__init__(envs)
-        self.parser = parser
 
-    def __delitem__(self, key):
-        super(Envs, self).__delitem__(key)
-        self.parser.envs = dict(self)
+class Args(KeyValues):
+    """
+    A class for allowing direct write access to Dockerfile build args, e.g.:
 
-    def __setitem__(self, key, value):
-        super(Envs, self).__setitem__(key, value)
-        self.parser.envs = dict(self)
+    parser.args['variable_name'] = 'value'
+    """
+    parser_attr = 'args'
 
 
 class DockerfileParser(object):
     def __init__(self, path=None,
                  cache_content=False,
                  env_replace=True,
                  parent_env=None,
+                 build_args=None,
```

Comment:
This has to be appended as last param, otherwise we may break BW compatibility for user who are just using it as positional args (I'd wish having `*` in params list in py2)